nimbus/src/evaluator.rs

```diff
@@ -28,65 +29,65 @@ impl Bucket {
-/// the user.
+/// - `experiment` - The experiment.
 ///
+/// An `ExperimentEnrollment` -  you need to inspect the EnrollmentStatus to
+/// determine if the user is actually enrolled.
 /// # Errors:
 ///
 /// The function can return errors in one of the following cases (but not limited to):
 ///
 /// - If the bucket sampling failed (i.e we could not find if the user should or should not be enrolled in the experiment based on the bucketing)
 /// - If an error occurs while determining the branch the user should be enrolled in any of the experiments
-#[allow(dead_code)]
-pub fn filter_enrolled(
+pub fn evaluate_enrollment(
     nimbus_id: &Uuid,
     available_randomization_units: &AvailableRandomizationUnits,
-    experiments: &[Experiment],
-) -> Result<Vec<EnrolledExperiment>> {
+    exp: &Experiment,
+) -> Result<ExperimentEnrollment> {
     let nimbus_id = nimbus_id.to_string();
-    let mut res = Vec::with_capacity(experiments.len());
-    for exp in experiments {
-        let bucket_config = exp.bucket_config.clone();
-        let id = match available_randomization_units
-            .get_value(&nimbus_id, &bucket_config.randomization_unit)
-        {
-            Some(id) => id,
-            None => {
-                // XXX: When we link we glean, it would be nice
-                // if we could emit a failure telemetry event here.
-                log::info!(
-                    "Could not find a suitable randomization unit for {}. Skipping experiment.",
-                    &exp.slug
-                );
-                continue;
-            }
-        };
-        if sampling::bucket_sample(
-            vec![id.to_owned(), bucket_config.namespace],
-            bucket_config.start,
-            bucket_config.count,
-            bucket_config.total,
-        )? {
-            res.push(EnrolledExperiment {
+    let bucket_config = exp.bucket_config.clone();
+    let id = match available_randomization_units
+        .get_value(&nimbus_id, &bucket_config.randomization_unit)
+    {
+        Some(id) => id,
+        None => {
+            // XXX: When we link we glean, it would be nice
+            // if we could emit a failure telemetry event here.
+            log::info!(
+                "Could not find a suitable randomization unit for {}. Skipping experiment.",
+                &exp.slug
+            );
+            return Ok(ExperimentEnrollment {
                 slug: exp.slug.clone(),
-                user_facing_name: exp.user_facing_name.clone(),
-                user_facing_description: exp.user_facing_description.clone(),
-                branch_slug: choose_branch(&exp.slug, &exp.branches, &id)?.clone().slug,
+                status: EnrollmentStatus::Error {
+                    reason: "No randomization unit".into(),
+                },
             });
         }
-    }
-    Ok(res)
+    };
+    let status = if sampling::bucket_sample(
```

Comment:
We are also not checking any filtering, but we should probably check that the `application` field in the experiment matches the `app_id` field in the `AppContext` at some point in here, in fact, it might be best to check that first to avoid the enrollment calculations if this isn't the intended app. Although, I suppose this isn't _really_ important until we have more than one application using nimbus on mobile...

Reply:
I wired up the existing `targeting` function, but didn't check the app field yet - we have a `struct Matcher` that, according to its comments, implies the server is going to send one of them in an experiment and we should be matching that way - but that contradicts with the server. So I think we need a new ticket on file to wire up "application matching" - and it that turns out to be literally just a string comparison, that's fine and we can remove the matcher.